File: frontend/src/api/testing.ts
import { apiConfig, request, simulateLatency } from "./client";
import { TESTING_STOCKS } from "@/mock/data/testingStocks";
import { RESEARCH_RESULTS } from "@/mock/data/researchResults";
import { MULTIBAGGER_CANDIDATES } from "@/mock/data/multibaggerCandidates";
import type { TrackedStock } from "@/types/testing";

export interface AddStockInput {
  ticker: string;
  company?: string;
  dateAdded: string;
  quantity?: number;
  /** if omitted, the backend/mock fetches the market price as of dateAdded */
  entryPrice?: number;
}

export interface UpdatePricesResult {
  updated: number;
  total: number;
  timestamp: string;
  prices: Record<string, number>;
}

function hashSeed(input: string): number {
  let h = 0;
  for (let i = 0; i < input.length; i++) h = (h * 31 + input.charCodeAt(i)) >>> 0;
  return h;
}

/** Looks up a plausible current price for a ticker. Prefers known mock
 * universe prices so tickers seen elsewhere in the demo stay consistent,
 * otherwise derives a stable pseudo-price from the ticker itself. */
function lookupPrice(ticker: string): { price: number; company: string } {
  const known = [...RESEARCH_RESULTS, ...MULTIBAGGER_CANDIDATES].find((x) => x.ticker.toUpperCase() === ticker.toUpperCase());
  if (known) return { price: known.price, company: known.company };
  const seed = hashSeed(ticker.toUpperCase());
  const price = 40 + (seed % 4600) / 10;
  return { price: Math.round(price * 100) / 100, company: ticker.toUpperCase() };
}

/**
 * Domain service for the Testing Lab.
 *
 *   GET  /testing/stocks
 *   POST /testing/stocks
 *   POST /testing/prices/update
 */
export const testingApi = {
  async listStocks(): Promise<TrackedStock[]> {
    if (apiConfig.useMocks) {
      await simulateLatency();
      return TESTING_STOCKS.map((s) => ({ ...s }));
    }
    return request<TrackedStock[]>("/testing/stocks");
  },

  async addStock(input: AddStockInput): Promise<TrackedStock> {
    if (apiConfig.useMocks) {
      await simulateLatency(300, 700);
      const looked = lookupPrice(input.ticker);
      const entryPrice = input.entryPrice ?? looked.price;
      const currentPrice = looked.price;
      const now = new Date().toISOString();
      return {
        id: `ts-${Date.now()}`,
        company: input.company?.trim() || looked.company,
        ticker: input.ticker.toUpperCase(),
        dateAdded: input.dateAdded,
        entryPrice,
        quantity: input.quantity,
        currentPrice,
        lastUpdated: now,
        status: "active",
        priceHistory: [
          { date: input.dateAdded, price: entryPrice },
          { date: now, price: currentPrice },
        ],
      };
    }
    return request<TrackedStock>("/testing/stocks", { method: "POST", body: JSON.stringify(input) });
  },

  async updatePrices(stocks: TrackedStock[]): Promise<UpdatePricesResult> {
    if (apiConfig.useMocks) {
      await simulateLatency(500, 1100);
      const prices: Record<string, number> = {};
      for (const s of stocks) {
        const drift = (Math.random() - 0.47) * 0.028;
        const next = Math.max(1, s.currentPrice * (1 + drift));
        prices[s.id] = Math.round(next * 100) / 100;
      }
      return { updated: stocks.length, total: stocks.length, timestamp: new Date().toISOString(), prices };
    }
    return request<UpdatePricesResult>("/testing/prices/update", {
      method: "POST",
      body: JSON.stringify({ ids: stocks.map((s) => s.id) }),
    });
  },
};

File: backend/app/api/routes/testing.py
import hashlib
import uuid
from datetime import datetime, timezone

from fastapi import APIRouter, Depends
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.models.tracked_stock import TrackedStock as TrackedStockModel
from app.schemas.testing import AddStockInput, TrackedStock, UpdatePricesInput, UpdatePricesResult

router = APIRouter(prefix="/testing", tags=["testing"])


def _to_schema(row: TrackedStockModel) -> TrackedStock:
    """Convert a TrackedStock ORM row into its API schema."""
    return TrackedStock.model_validate(row, from_attributes=True)


@router.get("/stocks", response_model=list[TrackedStock])
def list_stocks(db: Session = Depends(get_db)) -> list[TrackedStockModel]:
    """List every stock currently tracked in the Testing Lab."""
    return db.execute(select(TrackedStockModel)).scalars().all()


@router.post("/stocks", response_model=TrackedStock)
def add_stock(payload: AddStockInput, db: Session = Depends(get_db)) -> TrackedStockModel:
    """Add a new stock pick to the Testing Lab, using its entry price as the current price."""
    now = datetime.now(timezone.utc)
    entry_price = payload.entry_price if payload.entry_price is not None else _placeholder_price(payload.ticker)
    row = TrackedStockModel(
        id=f"ts-{uuid.uuid4().hex[:12]}",
        company=payload.company or payload.ticker.upper(),
        ticker=payload.ticker.upper(),
        date_added=payload.date_added,
        entry_price=entry_price,
        quantity=payload.quantity,
        current_price=entry_price,
        last_updated=now,
        price_history=[
            {"date": payload.date_added.isoformat(), "price": entry_price},
            {"date": now.isoformat(), "price": entry_price},
        ],
        status="active",
    )
    db.add(row)
    db.commit()
    db.refresh(row)
    return row


@router.post("/prices/update", response_model=UpdatePricesResult)
def update_prices(payload: UpdatePricesInput, db: Session = Depends(get_db)) -> UpdatePricesResult:
    """Refresh the current price for the given tracked stocks.

    Live market-data ingestion isn't wired up yet, so this keeps prices
    unchanged rather than fabricating a price movement.
    """
    now = datetime.now(timezone.utc)
    rows = db.execute(select(TrackedStockModel).where(TrackedStockModel.id.in_(payload.ids))).scalars().all()
    prices: dict[str, float] = {}
    for row in rows:
        row.last_updated = now
        prices[row.id] = row.current_price
    db.commit()
    return UpdatePricesResult(updated=len(rows), total=len(payload.ids), timestamp=now, prices=prices)


def _placeholder_price(ticker: str) -> float:
    """Derive a stable placeholder price from the ticker until live quotes are ingested."""
    seed = int(hashlib.sha256(ticker.upper().encode()).hexdigest(), 16)
    return round(40 + (seed % 4600) / 10, 2)
